src/textual/widgets/_tabs.py

```diff
@@ -0,0 +1,440 @@
+        ]
+        super().__init__(
+            name=name,
+            id=id,
+            classes=classes,
+            disabled=disabled,
+        )
+        self._tabs = add_tabs
+        self._first_active = active
+
+    @property
+    def _new_tab_id(self) -> int:
+        """Get the next tab id in a sequence."""
+        self._tabs_counter += 1
+        return self._tabs_counter
+
+    @property
+    def tab_count(self) -> int:
+        """The total number of tabs."""
+        return len(self.query("#tabs-list > Tab"))
+
+    @property
+    def _next_active(self) -> Tab | None:
+        """Get the next tab to make active, if the active tab is removed."""
+        tabs = list(self.query("#tabs-list > Tab").results(Tab))
+        if self.active_tab is None:
+            return None
+        try:
+            active_index = tabs.index(self.active_tab)
+        except ValueError:
+            return None
+        del tabs[active_index]
+        try:
+            return tabs[active_index]
+        except IndexError:
+            try:
+                return tabs[active_index - 1]
+            except IndexError:
+                pass
+        return None
+
+    def add_tab(self, tab: Tab | str | Text) -> None:
+        """Add a new tab to the end of the tab list.
+
+        Args:
+            tab: A new tab object, or a label (str or Text).
+        """
+        from_empty = self.tab_count == 0
+        tab_widget = (
+            Tab(tab, id=f"tab-{self._new_tab_id}")
+            if isinstance(tab, (str, Text))
+            else tab
+        )
+        mount_await = self.query_one("#tabs-list").mount(tab_widget)
+        if from_empty:
+            tab_widget.add_class("-active")
+            self.post_message(self.TabActivated(tab_widget))
+
+            async def refresh_active() -> None:
+                """Wait for things to be mounted before highlighting."""
+                await mount_await
+                self.active = tab_widget.id or ""
+                self._highlight_active(animate=False)
+
+            self.call_after_refresh(refresh_active)
+
+    def clear(self) -> None:
+        """Clear all the tabs."""
+        underline = self.query_one(Underline)
+        underline.highlight_start = 0
+        underline.highlight_end = 0
+        self.query("#tabs-list > Tab").remove()
+        self.post_message(self.TabActivated(None))
+
+    def remove_tab(self, tab_or_id: Tab | str | None) -> None:
+        """Remove a tab.
+
+        Args:
+            tab_id: The Tab's id.
+        """
+        if tab_or_id is None:
+            return
+        if isinstance(tab_or_id, Tab):
+            remove_tab = tab_or_id
+        else:
+            try:
+                remove_tab = self.query_one(f"#tabs-list > #{tab_or_id}", Tab)
+            except NoMatches:
+                return
+        removing_active_tab = remove_tab.has_class("-active")
+
+        next_tab = self._next_active
+        self.post_message(self.TabActivated(next_tab))
```

Comment:
Why do we do this outside of `do_remove`? Shouldn't all these actions be done together?

Reply:
The `do_remove` is mainly for the underline bar. So that it can move the start and end post-layout. This message doesn't need to wait, it can go straight away.